### PR TITLE
goflags support

### PR DESCRIPTION
**Pull request by LuitelSamikshya**
This PR has a dependency to  https://github.com/projectdiscovery/goflags/pull/32 and provides following support:

- provides go flags support

- safely remove following controls from the current config file as we already this information in the code `resolvers,sources, all-sources,recursive`
- Creates default goflags config for the subfinder at `$HOME/.config/subfinder/config.yaml`
- supports custom source/provider config file

### CODE REVIEW

**Review by Mzack9999 — CHANGES_REQUESTED** [review]
As discussed, if the config file can't be YAML unmarshaled due to deprecated structure, let's warn the user and exit. It may be possible that it's also necessary to modify https://github.com/projectdiscovery/goflags to detect the error. If that's the case, let's track it into a separate GH issue in the `goflags` repository.

**Comment by LuitelSamikshya** [comment]
[https://github.com/projectdiscovery/goflags/pull/32](url)

**Review by Mzack9999 — APPROVED** [review]

**Review by ehsandeep — COMMENTED** [review]
@LuitelSamikshya unable to see new config file and scan is not running with this branch as well, here is example outputs:

```console
$ git branch
  dev
* issue-378-goflag-support
  master
  pr/419
  pr/424
  pr/431
  pr/433
  pr/437
  pr/438
  pr/445
  pr/461

$ go build

$ ./subfinder -d hackerone.com

               __    _____           __         
   _______  __/ /_  / __(_)___  ____/ /__  _____
  / ___/ / / / __ \/ /_/ / __ \/ __  / _ \/ ___/
 (__  ) /_/ / /_/ / __/ / / / / /_/ /  __/ /    
/____/\__,_/_.___/_/ /_/_/ /_/\__,_/\___/_/ v2.4.9

		projectdiscovery.io

Use with caution. You are responsible for your actions
Developers assume no liability and are not responsible for any misuse or damage.
By using subfinder, you also agree to the terms of the APIs used.

[INF] loading the default
[INF] Enumerating subdomains for hackerone.com
[INF] Found 0 subdomains for hackerone.com in 144 microseconds

$ ls ~/.config/subfinder/
config.yaml
```

**Review by Mzack9999 — CHANGES_REQUESTED** [review]
It seems like there is some issue related to the integration - moving back to `in progress`:
```console
$ go run . -d hackerone.com

               __    _____           __         
   _______  __/ /_  / __(_)___  ____/ /__  _____
  / ___/ / / / __ \/ /_/ / __ \/ __  / _ \/ ___/
 (__  ) /_/ / /_/ / __/ / / / / /_/ /  __/ /    
/____/\__,_/_.___/_/ /_/_/ /_/\__,_/\___/_/ v2.4.9

                projectdiscovery.io

Use with caution. You are responsible for your actions
Developers assume no liability and are not responsible for any misuse or damage.
By using subfinder, you also agree to the terms of the APIs used.

[INF] loading the default
[INF] Enumerating subdomains for hackerone.com
[INF] Found 0 subdomains for hackerone.com in 258 microseconds
$ go build .
$ ./subfinder . -d hackerone.com

               __    _____           __         
   _______  __/ /_  / __(_)___  ____/ /__  _____
  / ___/ / / / __ \/ /_/ / __ \/ __  / _ \/ ___/
 (__  ) /_/ / /_/ / __/ / / / / /_/ /  __/ /    
/____/\__,_/_.___/_/ /_/_/ /_/\__,_/\___/_/ v2.4.9

                projectdiscovery.io

Use with caution. You are responsible for your actions
Developers assume no liability and are not responsible for any misuse or damage.
By using subfinder, you also agree to the terms of the APIs used.

[INF] loading the default
[FTL] Program exiting: no input list provided
$ ls ~/.config/subfinder/
config.json     config.yaml
```

**Review by ehsandeep — DISMISSED** [review]
Here are a few things that we need to consider for automatic migration of the existing provider keys/credentials to the new provider config file.

- Move existing **keys/token** of providers to the newly created provider config file at `$HOME/.config/subfinder/provider.yaml`
- New flag to use custom source/provider config file, as default `$HOME/.config/subfinder/provider.yaml` will be created and used.
- We can safely remove following controls from the current config file as we already this information in the code `resolvers`,`sources`, `all-sources`,`recursive`
- Create default `goflags` config for the subfinder at `$HOME/.config/subfinder/config.yaml`

**Comment by LuitelSamikshya** [comment]
[https://github.com/projectdiscovery/goflags/pull/32](url)

File: v2/pkg/runner/options.go [hunk 1]
```diff
@@ -104,15 +138,20 @@ func ParseOptions() *Options {
...
+	// Check if the application loading with any configuration, then take it
+	// Otherwise load the default config data from the code
+	if options.ConfigFile != "" {
+		gologger.Info().Msgf("loading from file %s", options.ConfigFile)
+		options.sourceRunTasks()
 	} else {
-		options.normalRunTasks()
+		gologger.Info().Msg("loading the default")
+		if flagSet.ProviderConfigFile {
+			options.ConfigFile = path.Join(config, "provider.yaml")
```
Comment (Mzack9999) [inline v2/pkg/runner/options.go:149]: please use `filepath.Join`

File: v2/pkg/runner/options.go [hunk 2]
```diff
@@ -104,15 +138,20 @@ func ParseOptions() *Options {
...
+		} else {
+			options.ConfigFile = path.Join(config, "config.yaml")
```
Comment (Mzack9999) [inline v2/pkg/runner/options.go:151]: please use `filepath.Join`

**Review by Mzack9999 — CHANGES_REQUESTED** [review]
It's necessary to update `go.mod` to use the latest commit of https://github.com/projectdiscovery/goflags/pull/32

File: v2/pkg/runner/options.go (hunk 1 above)
Comment (LuitelSamikshya) [reply v2/pkg/runner/options.go]: Done

File: v2/pkg/runner/options.go (hunk 2 above)
Comment (LuitelSamikshya) [reply v2/pkg/runner/options.go]: done

**Review by Mzack9999 — CHANGES_REQUESTED** [review]
Providers are correctly migrated to `$HOME/.config/subfinder/provider.yaml`, anyway invoking again the tool trigger an EOF error for `$HOME/.config/subfinder/config.yaml`. If no options are specified, the tool should load the default config from `config.yaml` and the providers from the default `provider.yaml`.

**Review by Mzack9999 — CHANGES_REQUESTED** [review]
If the `$home/.config/subfinder` folder is empty I guess the application should re-generate empty configuration files:
```console
$ rm -rf /Users/user/.config/subfinder/
$ echo hackerone.com | ./subfinder 

               __    _____           __         
   _______  __/ /_  / __(_)___  ____/ /__  _____
  / ___/ / / / __ \/ /_/ / __ \/ __  / _ \/ ___/
 (__  ) /_/ / /_/ / __/ / / / / /_/ /  __/ /    
/____/\__,_/_.___/_/ /_/_/ /_/\__,_/\___/_/ v2.4.9

                projectdiscovery.io

Use with caution. You are responsible for your actions
Developers assume no liability and are not responsible for any misuse or damage.
By using subfinder, you also agree to the terms of the APIs used.

[INF] loading the default
[FTL] Could not read configuration file /Users/user/.config/subfinder/provider.yaml: open /Users/user/.config/subfinder/provider.yaml: no such file or directory
```

**Comment by LuitelSamikshya** [comment]
Screen Shot 2022-02-07 at 10 42 02 AM

**Pull request closed without merging**